pkg/google/cloudsql/pricingmap.go

```diff
@@ -0,0 +1,433 @@
+	cpuPrice     float64 // per vCPU per hour if custom
+	ramPrice     float64 // per GB per hour if custom
+}
+
+type pricingMap struct {
+	logger    *slog.Logger
+	gcpClient client.Client
+	mu        sync.RWMutex
+	skus      []*billingpb.Sku
+}
+
+type instanceTraits struct {
+	region       string
+	dbType       string
+	availability string
+	spec         *instanceSpec
+}
+
+var (
+	ErrInstanceNotFound   = errors.New("instance not found")
+	ErrPriceNotFound      = errors.New("price not found for instance")
+	ErrInvalidTier        = errors.New("invalid tier format")
+	ErrRegionNotFound     = errors.New("region not found in pricing")
+	ErrCustomPriceMissing = errors.New("custom pricing (CPU/RAM) not available")
+
+	serviceName = "Cloud SQL"
+
+	// Tier parsing regex
+	tierCustomRegex = regexp.MustCompile(`^db-custom-(\d+)-(\d+)$`)
+	// Micro/small tier regex (e.g., db-f1-micro, db-g1-small). It has a different format than the generic regex.
+	tierMicroSmallRegex = regexp.MustCompile(`^db-([a-z0-9]+)-([a-z]+)$`)
+	// Generic regex that matches various db types and tier categories (e.g., standard, highmem, custom, etc.)
+	tierGenericRegex = regexp.MustCompile(`^db-([a-z0-9]+)-([a-z0-9]+)-(?:([a-z0-9]+)-)?(\d+)(?:-(\d+))?$`)
+)
+
+// Price calculation logic:
+// 1. If the instance tier matches the custom format (db-custom-{vCPU}-{RAM}), use custom pricing logic:
+//  - Find CPU and RAM component SKUs in the region (usage_unit "h" for CPU, "GiBy" for RAM)
+//  - Calculate total price: (vCPU count × CPU price per hour) + (RAM in GB × RAM price per GB per hour)
+// 2. If the instance is a standard instance (any other tier format), use standard pricing logic:
+//  - Find a SKU matching the region and tier type:
+//    * For micro/small tiers (db-f1-micro, db-g1-small): match by tier name in SKU description
+//    * For standard tiers (db-n1-standard-*, etc.): match by region only (first matching SKU)
+//  - Return the price from the matched SKU
+
+func newPricingMap(logger *slog.Logger, gcpClient client.Client) *pricingMap {
+	return &pricingMap{
+		logger:    logger,
+		gcpClient: gcpClient,
+		skus:      []*billingpb.Sku{},
+	}
+}
+
+func (pm *pricingMap) getSKus(ctx context.Context) error {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+
+	serviceFullName, err := pm.gcpClient.GetServiceName(ctx, serviceName)
+	if err != nil {
+		return fmt.Errorf("failed to get service name for Cloud SQL: %w", err)
+	}
+
+	skus := pm.gcpClient.GetPricing(ctx, serviceFullName)
+	if len(skus) == 0 {
+		pm.logger.Warn("no SKUs found for Cloud SQL service", "serviceName", serviceFullName)
+		return nil
+	}
+
+	pm.skus = skus
+	pm.logger.Info("loaded Cloud SQL pricing SKUs", "count", len(skus), "serviceName", serviceFullName)
+	return nil
+}
+
+func priceForSKU(sku *billingpb.Sku) (float64, bool) {
+	if sku == nil {
+		return 0, false
+	}
+	if len(sku.PricingInfo) == 0 {
+		return 0, false
+	}
+	if len(sku.PricingInfo[0].PricingExpression.TieredRates) == 0 {
+		return 0, false
+	}
```

Comment:
Is the price for the SKU really 0 or was it not found so we should be returning an error? 🤔